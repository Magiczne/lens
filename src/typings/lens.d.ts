import { LoadEvent, Viewport } from 'puppeteer'
import { Logger } from '@/typings/logging'
import { RulesetParser, RulesetValidator } from '@/typings/rules'

interface ArgumentParser {
    parse (args: LensArguments): ParsedLensArguments

    parseDirectory (dir?: string): string | undefined
    parseResolution (resolution?: string): Record<string, Viewport[]>
    parseUrl (url: string): URL[]
}

interface LensArguments {
    [x: string]: unknown

    url?: string
    resolution?: string
    tag: string

    inputDir?: string
    outputDir?: string

    _: string[]
    $0: string
}

interface ParsedLensArguments {
    urls: URL[]
    resolutions: Record<string, Viewport[]>
    tag: string

    inputDir?: string
    outputDir?: string
}

interface LensConfig {
    chunkSize: number
    directories: {
        input: string
        output: string
    }
    puppeteer: {
        headless: boolean
        waitUntil: LoadEvent | LoadEvent[]
    }
}

interface LensDependencies {
    argumentParser: ArgumentParser
    logger: Logger
    rulesetParser: RulesetParser
    rulesetValidator: RulesetValidator
}

export {
    ArgumentParser,
    LensArguments, ParsedLensArguments,
    LensConfig, LensDependencies
}
