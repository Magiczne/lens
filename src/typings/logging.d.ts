interface Logger {
    error (message: string): void
    header (message: string): void
    info (message: string): void
    success (message: string): void
}

export {
    Logger
}
