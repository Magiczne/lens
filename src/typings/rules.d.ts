import { Page, Viewport } from 'puppeteer'

interface Rule {
    url: string
    tag: string
    renderFor?: Array<string> | Array<Viewport>

    afterPageLoaded?: (page: Page) => Promise<void> | void
}

interface ParsedRule {
    url: URL
    tag: string
    renderFor: Record<string, Array<Viewport>>

    afterPageLoaded?: (page: Page) => Promise<void> | void
}

interface Ruleset {
    disable: boolean
    rules: Array<Rule>
}

interface ParsedRuleset {
    disable: boolean
    rules: Array<ParsedRule>
}

interface RulesetParser {
    parse (rawRuleset: Ruleset): ParsedRuleset

    parseRule (rule: Rule): ParsedRule
}

interface RulesetValidator {
    validate (ruleset: Record<string, unknown>, file: string): Promise<Ruleset> | Ruleset
}

export {
    Rule, Ruleset,
    ParsedRule, ParsedRuleset,
    RulesetParser, RulesetValidator
}
