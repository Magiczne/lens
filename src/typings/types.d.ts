import { UrlWithStringQuery } from 'url'

interface Resolution {
	width: number
	height: number
}

interface LensArguments {
	[x: string]: unknown

	url: string
	resolution?: string
	tag: string

	_: string[]
	$0: string
}

interface ParsedLensArguments {
	urls: UrlWithStringQuery[]
	resolutions?: Resolution[]
	tag: string
}

interface Logger {
	error (message: string): void
	header (message: string): void
	info (message: string): void
	success (message: string): void
}

export {
	LensArguments, ParsedLensArguments, Resolution,
	Logger
}
