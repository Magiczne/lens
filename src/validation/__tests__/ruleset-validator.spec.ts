import each from 'jest-each'

import { RulesetValidator } from '@/typings/rules'
import DefaultRulesetValidator from '@/validation/ruleset-validator'

describe('DefaultRulesetValidator', () => {
    let validator: RulesetValidator

    beforeEach(() => {
        validator = new DefaultRulesetValidator()
    })

    it('should return schema when input object is valid', async () => {
        const ruleset = {
            disable: false,
            rules: [{
                url: 'http://example.com',
                tag: 'tag',
                renderFor: [
                    { width: 1920, height: 1080 },
                    { width: 1280, height: 720 }
                ]
            }]
        }

        await expect(validator.validate(ruleset, 'ruleset.js'))
            .resolves
            .toStrictEqual(ruleset)
    })

    each([
        {
            rules: [{
                url: 'example.com',
                renderFor: [{ width: 1920, height: 1080 }]
            }]
        },
        {
            rules: [{
                url: 'ftp://example.com',
                renderFor: ['phone', 'tablet']
            }]
        },
        {
            rules: [{
                url: 'https://example.com'
            }]
        },
        {
            rules: [{
                url: 'https://example.com',
                renderFor: { foo: 'bar' }
            }]
        },
        {
            rules: [{
                url: 'https://example.com',
                async afterPageLoaded () {
                    return
                }
            }]
        }
    ]).it('should throw LensRulesetError when input is invalid', async (ruleset: Record<string, unknown>) => {
        await expect(validator.validate(ruleset, 'ruleset.js'))
            .rejects
            .toThrowErrorMatchingSnapshot()
    })
})
