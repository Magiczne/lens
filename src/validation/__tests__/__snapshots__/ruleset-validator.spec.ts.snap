// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultRulesetValidator should throw LensRulesetError when input is invalid 1`] = `"Ruleset ruleset.js contains errors"`;

exports[`DefaultRulesetValidator should throw LensRulesetError when input is invalid 2`] = `"Ruleset ruleset.js contains errors"`;

exports[`DefaultRulesetValidator should throw LensRulesetError when input is invalid 3`] = `"Ruleset ruleset.js contains errors"`;

exports[`DefaultRulesetValidator should throw LensRulesetError when input is invalid 4`] = `"Ruleset ruleset.js contains errors"`;

exports[`DefaultRulesetValidator should throw LensRulesetError when input is invalid 5`] = `"Ruleset ruleset.js contains errors"`;
